### PR TITLE
Validate host header

### PR DESCRIPTION
**Pull request by rnewson**
By default, the Host header must be either localhost or the
fully-qualified domain name of the host.

Administrators can disable this check with;

[httpd]
validate_host = false

Administrators can override the valid host list with;

[httpd]
valid_hosts = foo,bar,baz

COUCHDB-2752

### CODE REVIEW

**Comment by kxepal** [comment]
It has a little bit flaw: assume `valid_hosts = foo,bar,,baz` - small typo after removing some host. Normal situation. Then valid hosts list will be ["foo", "bar", [], "baz"]. If request will be send with no Host header, then `hostname/1` will return empty list `[]` which will bypass validation.

**Comment by rnewson** [comment]
I'll make it so Host header must exist if validate_hosts is true

**Comment by kxepal** [comment]
Exists and be not empty, yes (:

**Comment by rnewson** [comment]
If Host header is not sent, we look for, but can never find, the undefined atom in the list of valid hosts.

**Comment by kxepal** [comment]
Ah, good then. Funny, but this PR breaks dev/run script - guess why (:

**Comment by kxepal** [comment]
There CouchDB being requested by IP address 127.0.0.1 which is not in white default list.

File: src/couch_httpd.erl [hunk 1]
```diff
@@ -357,6 +359,38 @@ handle_request_int(MochiReq, DefaultFun,
...
+            undefined;
+        Host ->
+            [Name | _Port] = re:split(Host, ":", [{parts, 2}, {return, list}]),
+            Name
+    end.
+
+valid_hosts() ->
+    case config:get("httpd", "valid_hosts") of
+        undefined ->
+            ["localhost", net_adm:localhost()];
```
Comment (kxepal) [inline src/couch_httpd.erl]: I think worth to add both "127.0.0.1" and "::1" here.

Comment (kxepal) [inline src/couch_httpd.erl]: Oh, no. IPv6 will break regexp above to extract hostname /:

**Comment by rnewson** [comment]
@kxepal good catch on ip literals, I've updated. (though I note I use dev/run to test this and didn't hit a problem yet)

**Comment by kxepal** [comment]
Awesome! +1

**Pull request merged**